// @sparticuz/chromium/build/cjs/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// source/index.ts
var index_exports = {};
__export(index_exports, {
  default: () => index_default,
  inflate: () => inflate,
  setupLambdaEnvironment: () => setupLambdaEnvironment
});
module.exports = __toCommonJS(index_exports);
var import_node_fs3 = require("fs");
var import_node_os3 = require("os");
var import_node_path4 = require("path");

// source/helper.ts
var import_node_fs = require("fs");
var import_promises = require("fs/promises");
var import_node_os = require("os");
var import_node_path = require("path");
var import_node_stream = require("stream");
var import_promises2 = require("stream/promises");
var import_tar_fs = require("tar-fs");
var setupLambdaEnvironment = (baseLibPath) => {
  process.env["FONTCONFIG_PATH"] ??= (0, import_node_path.join)((0, import_node_os.tmpdir)(), "fonts");
  process.env["HOME"] ??= (0, import_node_os.tmpdir)();
  if (process.env["LD_LIBRARY_PATH"] === void 0) {
    process.env["LD_LIBRARY_PATH"] = baseLibPath;
  } else if (!process.env["LD_LIBRARY_PATH"].startsWith(baseLibPath)) {
    process.env["LD_LIBRARY_PATH"] = [
      baseLibPath,
      ...new Set(process.env["LD_LIBRARY_PATH"].split(":"))
    ].join(":");
  }
};
var isValidUrl = (input) => {
  try {
    const url = new URL(input);
    if (url.protocol === "https:") return true;
    if (url.protocol === "http:" && (url.hostname === "localhost" || url.hostname === "127.0.0.1" || url.hostname === "[::1]")) {
      return true;
    }
    return false;
  } catch {
    return false;
  }
};
var isRunningInAmazonLinux2023 = (nodeMajorVersion2) => {
  const awsExecEnv = process.env["AWS_EXECUTION_ENV"] ?? "";
  const awsLambdaJsRuntime = process.env["AWS_LAMBDA_JS_RUNTIME"] ?? "";
  const codebuildImage = process.env["CODEBUILD_BUILD_IMAGE"] ?? "";
  if (awsExecEnv.includes("20.x") || awsExecEnv.includes("22.x") || awsExecEnv.includes("24.x") || awsLambdaJsRuntime.includes("20.x") || awsLambdaJsRuntime.includes("22.x") || awsLambdaJsRuntime.includes("24.x") || codebuildImage.includes("nodejs20") || codebuildImage.includes("nodejs22") || codebuildImage.includes("nodejs24")) {
    return true;
  }
  if (process.env["VERCEL"] && nodeMajorVersion2 >= 20) {
    return true;
  }
  return false;
};
var downloadAndExtract = async (url) => {
  const destDir = (0, import_node_path.join)((0, import_node_os.tmpdir)(), "chromium-pack");
  const response = await fetch(url, {
    redirect: "follow",
    signal: AbortSignal.timeout(3e5)
  });
  if (!response.ok) {
    throw new Error(`Unexpected status code: ${String(response.status)}.`);
  }
  if (!response.body) {
    throw new Error("Response body is empty.");
  }
  try {
    await (0, import_promises2.pipeline)(
      import_node_stream.Readable.fromWeb(
        response.body
      ),
      (0, import_tar_fs.extract)(destDir)
    );
  } catch (error) {
    await new Promise((resolve) => {
      (0, import_node_fs.rm)(destDir, { force: true, recursive: true }, () => {
        resolve();
      });
    });
    throw error;
  }
  return destDir;
};

// source/lambdafs.ts
var import_node_fs2 = require("fs");
var import_node_os2 = require("os");
var import_node_path2 = require("path");
var import_node_zlib = require("zlib");
var import_tar_fs2 = require("tar-fs");
var inflate = (filePath) => {
  const output = filePath.includes("swiftshader") ? (0, import_node_os2.tmpdir)() : (0, import_node_path2.join)(
    (0, import_node_os2.tmpdir)(),
    (0, import_node_path2.basename)(filePath).replace(
      /\.(?:t(?:ar(?:\.(?:br|gz))?|br|gz)|br|gz)$/i,
      ""
    )
  );
  return new Promise((resolve, reject) => {
    if (filePath.includes("swiftshader")) {
      if ((0, import_node_fs2.existsSync)(`${output}/libGLESv2.so`)) {
        resolve(output);
        return;
      }
    } else if ((0, import_node_fs2.existsSync)(output)) {
      resolve(output);
      return;
    }
    const isBrotli = /br$/i.test(filePath);
    const isGzip = /gz$/i.test(filePath);
    const isTar = /\.t(?:ar(?:\.(?:br|gz))?|br|gz)$/i.test(filePath);
    const highWaterMark = 2 ** 22;
    const source = (0, import_node_fs2.createReadStream)(filePath, { highWaterMark });
    let target;
    const handleError = (error) => {
      reject(error);
    };
    source.once("error", handleError);
    if (isTar) {
      target = (0, import_tar_fs2.extract)(output);
      target.once("finish", () => {
        resolve(output);
      });
    } else {
      target = (0, import_node_fs2.createWriteStream)(output, { mode: 448 });
      target.once("close", () => {
        resolve(output);
      });
    }
    target.once("error", handleError);
    if (isBrotli || isGzip) {
      const decompressor = isBrotli ? (0, import_node_zlib.createBrotliDecompress)({ chunkSize: 2 ** 21 }) : (0, import_node_zlib.createUnzip)({ chunkSize: 2 ** 21 });
      decompressor.once("error", handleError);
      source.pipe(decompressor).pipe(target);
    } else {
      source.pipe(target);
    }
  });
};

// source/paths.cjs.ts
var import_node_path3 = require("path");
function getBinPath() {
  return (0, import_node_path3.join)((0, import_node_path3.dirname)(__filename), "..", "..", "bin");
}

// source/index.ts
var nodeMajorVersion = Number.parseInt(
  process.versions.node.split(".")[0] ?? ""
);
if (isRunningInAmazonLinux2023(nodeMajorVersion)) {
  setupLambdaEnvironment((0, import_node_path4.join)((0, import_node_os3.tmpdir)(), "al2023", "lib"));
}
var Chromium = class {
  /**
   * Returns a list of additional Chromium flags recommended for serverless environments.
   * The canonical list of flags can be found on https://peter.sh/experiments/chromium-command-line-switches/.
   * Most of below can be found here: https://github.com/GoogleChrome/chrome-launcher/blob/main/docs/chrome-flags-for-tools.md
   */
  static get args() {
    const chromiumFlags = [
      "--ash-no-nudges",
      // Avoids blue bubble "user education" nudges (eg., "… give your browser a new look", Memory Saver)
      "--disable-domain-reliability",
      // Disables Domain Reliability Monitoring, which tracks whether the browser has difficulty contacting Google-owned sites and uploads reports to Google.
      "--disable-print-preview",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kDisablePrintPreview&ss=chromium
      "--disk-cache-size=33554432",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kDiskCacheSize&ss=chromium Forces the maximum disk space to be used by the disk cache, in bytes.
      "--no-default-browser-check",
      // Disable the default browser check, do not prompt to set it as such. (This is already set by Playwright, but not Puppeteer)
      "--no-pings",
      // Don't send hyperlink auditing pings
      "--single-process",
      // Runs the renderer and plugins in the same process as the browser. NOTES: Needs to be single-process to avoid `prctl(PR_SET_NO_NEW_PRIVS) failed` error
      "--font-render-hinting=none"
      // https://github.com/puppeteer/puppeteer/issues/2410#issuecomment-560573612
    ];
    const chromiumDisableFeatures = [
      "AudioServiceOutOfProcess",
      "IsolateOrigins",
      "site-per-process"
      // Disables OOPIF. https://www.chromium.org/Home/chromium-security/site-isolation
    ];
    const chromiumEnableFeatures = ["SharedArrayBuffer"];
    const graphicsFlags = [
      "--ignore-gpu-blocklist",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kIgnoreGpuBlocklist&ss=chromium
      "--in-process-gpu"
      // Saves some memory by moving GPU process into a browser process thread
    ];
    if (this.graphics) {
      graphicsFlags.push(
        // As the unsafe WebGL fallback, SwANGLE (ANGLE + SwiftShader Vulkan)
        "--use-gl=angle",
        "--use-angle=swiftshader",
        "--enable-unsafe-swiftshader"
      );
    } else {
      graphicsFlags.push("--disable-webgl");
    }
    const insecureFlags = [
      "--allow-running-insecure-content",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kAllowRunningInsecureContent&ss=chromium
      "--disable-setuid-sandbox",
      // Lambda runs as root, so this is required to allow Chromium to run as root
      "--disable-site-isolation-trials",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kDisableSiteIsolation&ss=chromium
      "--disable-web-security"
      // https://source.chromium.org/search?q=lang:cpp+symbol:kDisableWebSecurity&ss=chromium
    ];
    const headlessFlags = [
      "--headless='shell'",
      // We only support running chrome-headless-shell
      "--no-sandbox",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kNoSandbox&ss=chromium
      "--no-zygote"
      // https://source.chromium.org/search?q=lang:cpp+symbol:kNoZygote&ss=chromium
    ];
    return [
      ...chromiumFlags,
      `--disable-features=${[...chromiumDisableFeatures].join(",")}`,
      `--enable-features=${[...chromiumEnableFeatures].join(",")}`,
      ...graphicsFlags,
      ...insecureFlags,
      ...headlessFlags
    ];
  }
  /**
   * Returns whether the graphics stack is enabled or disabled
   * @returns boolean
   */
  static get graphics() {
    return this.graphicsMode;
  }
  /**
   * Sets whether the graphics stack is enabled or disabled.
   * @param true means the stack is enabled. WebGL will work.
   * @param false means that the stack is disabled. WebGL will not work.
   * @default true
   */
  static set setGraphicsMode(value) {
    if (typeof value !== "boolean") {
      throw new TypeError(
        `Graphics mode must be a boolean, you entered '${String(value)}'`
      );
    }
    this.graphicsMode = value;
  }
  /**
   * If true, the graphics stack and webgl is enabled,
   * If false, webgl will be disabled.
   * (If false, the swiftshader.tar.br file will also not extract)
   */
  static graphicsMode = true;
  /**
   * Inflates the included version of Chromium
   * @param input The location of the `bin` folder
   * @returns The path to the `chromium` binary
   */
  static async executablePath(input) {
    if ((0, import_node_fs3.existsSync)((0, import_node_path4.join)((0, import_node_os3.tmpdir)(), "chromium"))) {
      return (0, import_node_path4.join)((0, import_node_os3.tmpdir)(), "chromium");
    }
    if (input && isValidUrl(input)) {
      return this.executablePath(await downloadAndExtract(input));
    }
    input ??= getBinPath();
    if (!(0, import_node_fs3.existsSync)(input)) {
      throw new Error(
        `The input directory "${input}" does not exist. If you are using a bundler (esbuild, webpack, etc.), you must externalize @sparticuz/chromium so it is not relocated. See: https://github.com/Sparticuz/chromium#bundler-configuration`
      );
    }
    const promises = [
      inflate((0, import_node_path4.join)(input, "chromium.br")),
      inflate((0, import_node_path4.join)(input, "fonts.tar.br")),
      inflate((0, import_node_path4.join)(input, "swiftshader.tar.br"))
    ];
    if (isRunningInAmazonLinux2023(nodeMajorVersion)) {
      promises.push(inflate((0, import_node_path4.join)(input, "al2023.tar.br")));
    }
    const result = await Promise.all(promises);
    return result.shift();
  }
};
var index_default = Chromium;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  inflate,
  setupLambdaEnvironment
});
module.exports = module.exports.default;
