// @sparticuz/chromium/build/esm/index.js
// source/index.ts
import { existsSync as existsSync2 } from "fs";
import { tmpdir as tmpdir3 } from "os";
import { join as join4 } from "path";

// source/helper.ts
import { createWriteStream, rm } from "fs";
import { access, symlink } from "fs/promises";
import { tmpdir } from "os";
import { join } from "path";
import { Readable } from "stream";
import { pipeline } from "stream/promises";
import { extract } from "tar-fs";
var setupLambdaEnvironment = (baseLibPath) => {
  process.env["FONTCONFIG_PATH"] ??= join(tmpdir(), "fonts");
  process.env["HOME"] ??= tmpdir();
  if (process.env["LD_LIBRARY_PATH"] === void 0) {
    process.env["LD_LIBRARY_PATH"] = baseLibPath;
  } else if (!process.env["LD_LIBRARY_PATH"].startsWith(baseLibPath)) {
    process.env["LD_LIBRARY_PATH"] = [
      baseLibPath,
      ...new Set(process.env["LD_LIBRARY_PATH"].split(":"))
    ].join(":");
  }
};
var isValidUrl = (input) => {
  try {
    const url = new URL(input);
    if (url.protocol === "https:") return true;
    if (url.protocol === "http:" && (url.hostname === "localhost" || url.hostname === "127.0.0.1" || url.hostname === "[::1]")) {
      return true;
    }
    return false;
  } catch {
    return false;
  }
};
var isRunningInAmazonLinux2023 = (nodeMajorVersion2) => {
  const awsExecEnv = process.env["AWS_EXECUTION_ENV"] ?? "";
  const awsLambdaJsRuntime = process.env["AWS_LAMBDA_JS_RUNTIME"] ?? "";
  const codebuildImage = process.env["CODEBUILD_BUILD_IMAGE"] ?? "";
  if (awsExecEnv.includes("20.x") || awsExecEnv.includes("22.x") || awsExecEnv.includes("24.x") || awsLambdaJsRuntime.includes("20.x") || awsLambdaJsRuntime.includes("22.x") || awsLambdaJsRuntime.includes("24.x") || codebuildImage.includes("nodejs20") || codebuildImage.includes("nodejs22") || codebuildImage.includes("nodejs24")) {
    return true;
  }
  if (process.env["VERCEL"] && nodeMajorVersion2 >= 20) {
    return true;
  }
  return false;
};
var downloadAndExtract = async (url) => {
  const destDir = join(tmpdir(), "chromium-pack");
  const response = await fetch(url, {
    redirect: "follow",
    signal: AbortSignal.timeout(3e5)
  });
  if (!response.ok) {
    throw new Error(`Unexpected status code: ${String(response.status)}.`);
  }
  if (!response.body) {
    throw new Error("Response body is empty.");
  }
  try {
    await pipeline(
      Readable.fromWeb(
        response.body
      ),
      extract(destDir)
    );
  } catch (error) {
    await new Promise((resolve) => {
      rm(destDir, { force: true, recursive: true }, () => {
        resolve();
      });
    });
    throw error;
  }
  return destDir;
};

// source/lambdafs.ts
import { createReadStream, createWriteStream as createWriteStream2, existsSync } from "fs";
import { tmpdir as tmpdir2 } from "os";
import { basename, join as join2 } from "path";
import { createBrotliDecompress, createUnzip } from "zlib";
import { extract as extract2 } from "tar-fs";
var inflate = (filePath) => {
  const output = filePath.includes("swiftshader") ? tmpdir2() : join2(
    tmpdir2(),
    basename(filePath).replace(
      /\.(?:t(?:ar(?:\.(?:br|gz))?|br|gz)|br|gz)$/i,
      ""
    )
  );
  return new Promise((resolve, reject) => {
    if (filePath.includes("swiftshader")) {
      if (existsSync(`${output}/libGLESv2.so`)) {
        resolve(output);
        return;
      }
    } else if (existsSync(output)) {
      resolve(output);
      return;
    }
    const isBrotli = /br$/i.test(filePath);
    const isGzip = /gz$/i.test(filePath);
    const isTar = /\.t(?:ar(?:\.(?:br|gz))?|br|gz)$/i.test(filePath);
    const highWaterMark = 2 ** 22;
    const source = createReadStream(filePath, { highWaterMark });
    let target;
    const handleError = (error) => {
      reject(error);
    };
    source.once("error", handleError);
    if (isTar) {
      target = extract2(output);
      target.once("finish", () => {
        resolve(output);
      });
    } else {
      target = createWriteStream2(output, { mode: 448 });
      target.once("close", () => {
        resolve(output);
      });
    }
    target.once("error", handleError);
    if (isBrotli || isGzip) {
      const decompressor = isBrotli ? createBrotliDecompress({ chunkSize: 2 ** 21 }) : createUnzip({ chunkSize: 2 ** 21 });
      decompressor.once("error", handleError);
      source.pipe(decompressor).pipe(target);
    } else {
      source.pipe(target);
    }
  });
};

// source/paths.ts
import { dirname, join as join3 } from "path";
import { fileURLToPath } from "url";
function getBinPath() {
  return join3(dirname(fileURLToPath(import.meta.url)), "..", "..", "bin");
}

// source/index.ts
var nodeMajorVersion = Number.parseInt(
  process.versions.node.split(".")[0] ?? ""
);
if (isRunningInAmazonLinux2023(nodeMajorVersion)) {
  setupLambdaEnvironment(join4(tmpdir3(), "al2023", "lib"));
}
var Chromium = class {
  /**
   * Returns a list of additional Chromium flags recommended for serverless environments.
   * The canonical list of flags can be found on https://peter.sh/experiments/chromium-command-line-switches/.
   * Most of below can be found here: https://github.com/GoogleChrome/chrome-launcher/blob/main/docs/chrome-flags-for-tools.md
   */
  static get args() {
    const chromiumFlags = [
      "--ash-no-nudges",
      // Avoids blue bubble "user education" nudges (eg., "… give your browser a new look", Memory Saver)
      "--disable-domain-reliability",
      // Disables Domain Reliability Monitoring, which tracks whether the browser has difficulty contacting Google-owned sites and uploads reports to Google.
      "--disable-print-preview",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kDisablePrintPreview&ss=chromium
      "--disk-cache-size=33554432",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kDiskCacheSize&ss=chromium Forces the maximum disk space to be used by the disk cache, in bytes.
      "--no-default-browser-check",
      // Disable the default browser check, do not prompt to set it as such. (This is already set by Playwright, but not Puppeteer)
      "--no-pings",
      // Don't send hyperlink auditing pings
      "--single-process",
      // Runs the renderer and plugins in the same process as the browser. NOTES: Needs to be single-process to avoid `prctl(PR_SET_NO_NEW_PRIVS) failed` error
      "--font-render-hinting=none"
      // https://github.com/puppeteer/puppeteer/issues/2410#issuecomment-560573612
    ];
    const chromiumDisableFeatures = [
      "AudioServiceOutOfProcess",
      "IsolateOrigins",
      "site-per-process"
      // Disables OOPIF. https://www.chromium.org/Home/chromium-security/site-isolation
    ];
    const chromiumEnableFeatures = ["SharedArrayBuffer"];
    const graphicsFlags = [
      "--ignore-gpu-blocklist",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kIgnoreGpuBlocklist&ss=chromium
      "--in-process-gpu"
      // Saves some memory by moving GPU process into a browser process thread
    ];
    if (this.graphics) {
      graphicsFlags.push(
        // As the unsafe WebGL fallback, SwANGLE (ANGLE + SwiftShader Vulkan)
        "--use-gl=angle",
        "--use-angle=swiftshader",
        "--enable-unsafe-swiftshader"
      );
    } else {
      graphicsFlags.push("--disable-webgl");
    }
    const insecureFlags = [
      "--allow-running-insecure-content",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kAllowRunningInsecureContent&ss=chromium
      "--disable-setuid-sandbox",
      // Lambda runs as root, so this is required to allow Chromium to run as root
      "--disable-site-isolation-trials",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kDisableSiteIsolation&ss=chromium
      "--disable-web-security"
      // https://source.chromium.org/search?q=lang:cpp+symbol:kDisableWebSecurity&ss=chromium
    ];
    const headlessFlags = [
      "--headless='shell'",
      // We only support running chrome-headless-shell
      "--no-sandbox",
      // https://source.chromium.org/search?q=lang:cpp+symbol:kNoSandbox&ss=chromium
      "--no-zygote"
      // https://source.chromium.org/search?q=lang:cpp+symbol:kNoZygote&ss=chromium
    ];
    return [
      ...chromiumFlags,
      `--disable-features=${[...chromiumDisableFeatures].join(",")}`,
      `--enable-features=${[...chromiumEnableFeatures].join(",")}`,
      ...graphicsFlags,
      ...insecureFlags,
      ...headlessFlags
    ];
  }
  /**
   * Returns whether the graphics stack is enabled or disabled
   * @returns boolean
   */
  static get graphics() {
    return this.graphicsMode;
  }
  /**
   * Sets whether the graphics stack is enabled or disabled.
   * @param true means the stack is enabled. WebGL will work.
   * @param false means that the stack is disabled. WebGL will not work.
   * @default true
   */
  static set setGraphicsMode(value) {
    if (typeof value !== "boolean") {
      throw new TypeError(
        `Graphics mode must be a boolean, you entered '${String(value)}'`
      );
    }
    this.graphicsMode = value;
  }
  /**
   * If true, the graphics stack and webgl is enabled,
   * If false, webgl will be disabled.
   * (If false, the swiftshader.tar.br file will also not extract)
   */
  static graphicsMode = true;
  /**
   * Inflates the included version of Chromium
   * @param input The location of the `bin` folder
   * @returns The path to the `chromium` binary
   */
  static async executablePath(input) {
    if (existsSync2(join4(tmpdir3(), "chromium"))) {
      return join4(tmpdir3(), "chromium");
    }
    if (input && isValidUrl(input)) {
      return this.executablePath(await downloadAndExtract(input));
    }
    input ??= getBinPath();
    if (!existsSync2(input)) {
      throw new Error(
        `The input directory "${input}" does not exist. If you are using a bundler (esbuild, webpack, etc.), you must externalize @sparticuz/chromium so it is not relocated. See: https://github.com/Sparticuz/chromium#bundler-configuration`
      );
    }
    const promises = [
      inflate(join4(input, "chromium.br")),
      inflate(join4(input, "fonts.tar.br")),
      inflate(join4(input, "swiftshader.tar.br"))
    ];
    if (isRunningInAmazonLinux2023(nodeMajorVersion)) {
      promises.push(inflate(join4(input, "al2023.tar.br")));
    }
    const result = await Promise.all(promises);
    return result.shift();
  }
};
var index_default = Chromium;
export {
  index_default as default,
  inflate,
  setupLambdaEnvironment
};
